metrics: add `read_distribution` histogram

WillemKauf opened the pull request:
Adds a `read_distribution` histogram to the `partition_probe`, which tracks the timestamp delta between data read by Kafka fetches in `read_from_partition()` and the current tip of the log.

This is a new histogram type, `log_hist_read_dist`. The timestamp delta is measured in minutes using 16 buckets in the histogram, such that the bounds span from 1 minute in the first bucket to 91 days in the last bucket.

A new boolean `cumulative` is also added to the template arguments in `logform_config`, to enable creation of both cumulative and non-cumulative histograms. It defaults to `true` for all existing histograms.

There are no updates made to the Redpanda dashboard yet, so it is up to the user to configure a panel with this histogram if they are interested in seeing the statistics.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.1.x
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

### Features

* Adds a read distribution histogram that can be visualized using Prometheus and Grafana:
  * redpanda_kafka_read_distribution_bucket
  * redpanda_kafka_read_distribution_count
  * redpanda_kafka_read_distribution_sum

andrwng reviewed (commented):
Structurally looks pretty good, just a couple high level questions about the max timestamp.

Also it'd be great to include some tests (I think fixture tests might be able to directly access the partition probe? otherwise a single-node test in ducktape is fine)

WillemKauf commented on src/v/cluster/partition_probe.h:
```diff
@@ -32,6 +33,7 @@ class partition_probe {
         virtual void add_bytes_fetched_from_follower(uint64_t) = 0;
         virtual void add_schema_id_validation_failed() = 0;
         virtual void setup_metrics(const model::ntp&) = 0;
+        virtual void add_read_event_delta_from_tip(uint64_t delta_from_tip) = 0;
```
Would love to do a templated function here, but due to PIMPL and virtual functions I cannot- I'll add the `minutes` suffix.

WillemKauf commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -116,6 +116,16 @@ static ss::future<read_result> read_from_partition(
         part.probe().add_bytes_fetched(data->size_bytes());
+
+        if (data->size_bytes() > 0) {
+            auto delta_from_tip_ms = part.max_data_timestamp()
+                                     - result.first_timestamp;
```
Agreed- if we handle this with `if (max_timestamp >= result.first_timestamp)`, I think that should be enough without getting into `std::optional<>`.

WillemKauf commented on src/v/kafka/server/replicated_partition.cc:
```diff
@@ -646,4 +646,8 @@ result<partition_info> replicated_partition::get_partition_info() const {
 }
 
+model::timestamp replicated_partition::max_data_timestamp() const {
+    return _partition->max_data_timestamp();
```
Great question- if the goal was to be able to study read patterns in order to adjust local retention values, measuring against `now()` would produce delta values that are more in line with "real time", and closer to the way local retention produces candidates for garbage collection. Will align with @piyushredpanda on this.

piyushredpanda replied:
yeah, great question, thanks for raising. I do agree that now() makes more sense.

WillemKauf commented:
> Structurally looks pretty good, just a couple high level questions about the max timestamp.
> 
> Also it'd be great to include some tests (I think fixture tests might be able to directly access the partition probe? otherwise a single-node test in ducktape is fine)

Added ducktape test `test_read_distribution_metric`, removed `max_timestamp()` related functions, and we now compare the timestamp delta against `model::timestamp::now()` for a more real-time age of data fetched.

andrwng commented on tests/rptest/tests/read_distribution_test.py:
```diff
@@ -0,0 +1,85 @@
+        self.timespan_ms = 24 * 60 * 60 * 1000  #One day in ms.
+        self.timestamp_step_ms = self.timespan_ms // (self.record_count - 1)
+        self.curr_ts = round(time.time() * 1000)
+        self.base_ts = self.curr_ts - self.timespan_ms  #Start producing from record_count days ago.
```
nit: seems like this is just one day ago?

WillemKauf replied:
Oops, outdated comment. Removed.

andrwng commented on tests/rptest/tests/read_distribution_test.py:
```diff
@@ -0,0 +1,85 @@
+        cumulative_topic_read_dist = [
+            (s.value, s.labels['le']) for s in read_dist_metric.samples if
+            s.labels['redpanda_topic'] == self.topic_name and 'le' in s.labels
+        ]
```
nit: maybe worth checking that we have the right number of buckets in the histogram? Otherwise it seems like a bug could cause the distribution to be empty and we wouldn't notice

WillemKauf replied:
Agreed!

andrwng commented on tests/rptest/tests/read_distribution_test.py:
```diff
@@ -0,0 +1,85 @@
+        expected_read_dist = {bucket: 0 for bucket in topic_read_dist.keys()}
+        bucket_list = list(topic_read_dist.keys())
+        for i in range(0, self.record_count):
+            t_minutes = round(
+                (self.curr_ts -
+                 (self.base_ts + i * self.timestamp_step_ms)) / (1000 * 60))
+            bucket = bucket_list[bisect.bisect_left(bucket_list, t_minutes)]
```
nit: mind adding a comment explaining what we're doing here? On first read this is hard to follow

andrwng replied:
Also, it kind of seems like it should be a fixed distribution since base_ts is curr_ts - timespan_ms. Given the bucketing will likely not change much, do you think it'd make sense to build the map manually with fixed values? IMO it'd be much easier to verify that the distribution is what we expect

WillemKauf replied:
Hmm. This was actually my attempt to stay away from hardcoding any map values, just in case any of the test parameters or histogram bucketing values are changed later.

I think I prefer this way of checking the distribution, but if you feel strongly that we should build the map manually, I can make the change.

Added a comment for now explaining the `bisect_left` portion.

andrwng replied:
In principal I agree with you, it makes the code adaptable to further bucketing/test changes. OTOH it introduces room for bugs in this test code. I'm fine with it being here, but as a reader I think it would help to add some hardcoded values (i.e maybe let's do both)

WillemKauf replied:
Added some hardcoded values to the `read_distribution_test`.

WillemKauf commented:
/ci-repeat

StephanDollberg commented on src/v/cluster/partition_probe.cc:
```diff
@@ -290,6 +290,13 @@ void replicated_partition_probe::setup_public_metrics(const model::ntp& ntp) {
+          [this] { return _read_distribution.read_dist_histogram_logform(); },
+          sm::description("Read path time distribution histogram"),
+          labels)
+          .aggregate({sm::shard_label, partition_label}),
```
I think we need to also aggregate the topic here. Otherwise public metrics will massively explode for many-topic clusters.

Side question, is there a reason we put this in public metrics?

WillemKauf replied:
I think that per-topic read distributions are probably a better source of information than one aggregated histogram, but I do agree that the histogram should be moved to internal metrics. Will fix.

StephanDollberg reviewed (commented):
I am wondering whether we somehow want to weigh this by bytes?

Like this we don't differentiate between a 100 byte read and a 100KB read.

Maybe this makes no difference because both kinda have the same overhead when having to reach out to cloud and I guess the main thing why we are interested in this is for local retention reasons.

WillemKauf commented:
Thanks for the extensive feedback @StephanDollberg. 

- Reworked the probe into the `kafka::server` object
- The metric is no longer associated with any NTP aggregation, but is still aggregated across all shards (by default).

ballard26 commented on src/v/utils/log_hist.h:
```diff
@@ -199,6 +199,8 @@ class log_hist {
      */
     seastar::metrics::histogram internal_histogram_logform() const;
 
+    seastar::metrics::histogram read_dist_histogram_logform() const;
```
`log_hist` is designed to be an approximation of a histogram. Take the type;
```
using log_hist_internal = log_hist<std::chrono::microseconds, 26, 8ul>;
```
there will be 26 buckets each of which is an unsigned integer. Each bucket will end up as another metric series. So 26 in this case or 18 in the public histogram case.  for `log_hist_internal` the buckets are laid out like; 
```
[0, 8us), [8us, 16us), [16us, 32us), [32us, 64us)....[~1min, +inf)
```
when a duration is record the bucket whose range includes the duration is found. Then the int for that bucket is incremented.

From this we can see that the two existing resuables are great for recording latency since there is a lot of granularity around the 0 to 1s range. However, for this metric it seems we care a bit more about broader ranges like minutes or hours. So the existing resuables won't work as all those values will be recorded to the last bucket and the resulting histogram won't be very useful.

So to that ends I think what @WillemKauf has done here is fine. It gives up the granularity we have in the latency histograms around 0-1s in exchange for having more buckets over a longer range of time.

ballard26 replied:
We probably want to make the comments on the resuable types a bit more instructive. I.e, this is the histogram you want to use if you only care about accurately representing sub-second durations, minute/hour durations, day durations... etc.

ballard26 replied:
TDLR; the resuable types were only designed for sub-second latency durations. It makes sense to me to add more that more accurately represent other ranges of times folks may be interested in.

StephanDollberg commented on src/v/kafka/read_distribution_probe.h:
```diff
@@ -0,0 +1,66 @@
+              },
+              sm::description("Read path time distribution histogram"),
+              {})
+              .aggregate({sm::shard_label}),
```
You want the version which you posted to me yesterday where you pass the aggregation labels to `add_group`. This allows for aggregation to change at runtime without having to restart.

Any reason you switched it?

StephanDollberg commented on src/v/kafka/read_distribution_probe.h:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        _metrics.add_group(
+          prometheus_sanitize::metrics_name("kafka:fetch"),
```
While `:` is a valid metric name character `prometheus_sanitize::metrics_name` immediately changes it `_` so I don't think there is much point in having it being a colon in the first place.

(Know we do it in other places as well but always found it odd).

WillemKauf replied:
Changed `:` to `_`.
